### DISCUSSION TITLE
Can generators run in-place?

### DISCUSSION THREAD

**Discussion by popizdeh** [halide/Halide#6324] [Q&A] [answered]
What would happen if both `Input<Buffer>` and `Output<Buffer` were pointing to the same memory? I have a feeling this is not safe...

**Comment by peteraarts811** [comment, marked as answer]
Currently we actually do this.
However, you need to take this situation into account when creating your schedule.
Currently our solution is somewhat like the following:

1. intermediate_buffers = some_function1(input_buffers)
2. output_buffers = some_function2(intermediate_buffers )

In our schedule we do a compute_root on intermediate_buffers,
after which it is safe to write the output_buffers (which will actually overwrite the input_buffers).

You do need to take into account though that the auto-scheduler will probably not help you (as far as I know..).

**Reply by popizdeh** [reply]
Thanks @peteraarts811 but is this truly in-place? Doesn't `compute_root()` on intermediate buffers allocate another image sized buffer?

**Reply by steven-johnson** [reply]
Depends on how you define 'in-place': if that means "the input buffer and output buffer can point to the same memory", then yes. If you *also* mean "there is no memory allocation done for any intermediate computation", then no. 

It's worth pointing out that achieving the latter is challenging regardless of the coding language you use; one can certainly write trivial examples that can get by with only register usage, but that's unlikely to be practical for a huge swath of algorithms.

**Reply by popizdeh** [reply]
Thanks for the clarification @steven-johnson, I was hoping to get the latter but was very much expecting to get the answer you provided.

To give you a bit more information I'm dealing with edge blur where edge pixels are just a fraction of the whole image, so reading and writing each pixel guarantees poor performance regardless of how you schedule it. But I understand this type of algorithm is not the common case for Halide.
